marlowe-runtime/indexer/Language/Marlowe/Runtime/Indexer/Types.hs

```diff
@@ -0,0 +1,366 @@
+
+  , txOutRef :: TxOutRef
+  -- ^ The unspent output.
+
+  , marloweAddress :: Address
+  -- ^ The address of the marlowe validator.
+
+  , payoutValidatorHash :: ScriptHash
+  -- ^ The hash of the payout validator.
+  } deriving Generic
+
+instance ToJSON UnspentContractOutput
+
+instance Eq UnspentContractOutput where
+  UnspentContractOutput (Core.SomeMarloweVersion Core.MarloweV1) refA addrA pHashA
+    == UnspentContractOutput (Core.SomeMarloweVersion Core.MarloweV1) refB addrB pHashB =
+    refA == refB && addrA == addrB && pHashA == pHashB
+
+
+instance Show UnspentContractOutput where
+  showsPrec p (UnspentContractOutput (Core.SomeMarloweVersion Core.MarloweV1) ref addr pHash) =
+    showParen (p >= 11)
+      ( showString "UnspentContractOutput"
+      . showSpace
+      . showsPrec 11 Core.MarloweV1
+      . showSpace
+      . showsPrec 11 ref
+      . showSpace
+      . showsPrec 11 addr
+      . showSpace
+      . showsPrec 11 pHash
+      )
+
+
+-- Extracts a MarloweBlock from Cardano Block information. Returns an updated
+-- MarloweUTxO.
+extractMarloweBlock
+  :: SystemStart
+  -> EraHistory CardanoMode
+  -> Set ScriptHash
+  -- ^ All known Marlowe script hashes.
+  -> BlockHeader
+  -- ^ The BlockHeader of the block.
+  -> Set Transaction
+  -- ^ The set of transactions in the block.
+  -> MarloweUTxO
+  -- ^ The current MarloweUTxO
+  -> Maybe (MarloweUTxO, MarloweBlock)
+extractMarloweBlock systemStart eraHistory marloweScriptHashes blockHeader txs =
+  sequenceA . swap . runState do
+    transactions <- execWriterT $ retrySilentUntilAllSilent (Set.toList txs) \tx -> do
+      extractCreateTx marloweScriptHashes tx
+      extractApplyInputsTx systemStart eraHistory blockHeader tx
+      extractWithdrawTx tx
+    pure case transactions of
+      [] -> Nothing
+      x : xs -> Just MarloweBlock { blockHeader, transactions = x :| xs }
+
+-- Runs the given writer action for each element in a structure. If any actions
+-- produce output, the silent elements
+-- (the ones that did not produce any output) will be retried
+retrySilentUntilAllSilent :: (Witherable t, MonadWriter [w] m) => t a -> (a -> m ()) -> m ()
+retrySilentUntilAllSilent as f = do
+  (as', allSilent) <- listens null $ flip wither as \a -> do
+    (_, isSilent) <- listens null $ f a
+    pure $ a <$ guard isSilent
+  if allSilent then pure () else retrySilentUntilAllSilent as' f
+
+-- | Extracts a MarloweCreateTransaction from a Chain transaction. A single
+-- transaction can create multiple Marlowe contracts, and this function returns
+-- a map of outputs that it failed to extract as well as the map of contracts
+-- it successfully extracted.
+extractCreateTx
+  :: Set ScriptHash
+  -- ^ All known Marlowe script hashes.
+
+  -> Transaction
+  -> WriterT [MarloweTransaction] (State MarloweUTxO) ()
+extractCreateTx marloweScriptHashes Transaction{..} = do
+  -- Creation transactions cannot consume outputs from other Marlowe contracts.
```

Comment:
This might be an issue if we later change the validator to allow a Marlowe contract to close in the same transaction that creates another Marlowe contract.

Reply:
I know - I shudder to think about that. For now though, it's explicitly disallowed.